File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Java/MergeTwoSortedLists.java
import java.util.*;


// Definition for singly-linked list.
public class MergeTwoSortedLists {
  public static class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
  }

  public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
    ListNode sentry = new ListNode(0);
    ListNode current = sentry;

    while(l1 != null && l2 != null) {
      if(l1.val <= l2.val){
        current.next = l1;
        current = l1;
        l1 = l1.next;
      } else {
        current.next = l2;
        current = l2;
        l2 = l2.next;
      }
    }

    current.next = l1 == null ? l2 : l1;

    return sentry.next;
  }

  public static void main(String[] args) {
    MergeTwoSortedLists mergeLists = new MergeTwoSortedLists();

    ListNode l1 = null;
    ListNode l2 = new ListNode(0);

    ListNode newHead = mergeLists.mergeTwoLists(l1, l2);

    while( newHead != null ){
      System.out.println(newHead);
      newHead = newHead.next;
    }

  }
}

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/MergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        ListNode * head, * temp;
        
        if( l1 == NULL )
            return l2;  // l2 may be NULL
        if( l2 == NULL )
            return l1;
         
        // Set the head and forward node
        if( l1->val <= l2->val){
            head=l1;
            l1=l1->next;
        }  
        else{
            head=l2;
            l2=l2->next;
        }
        
        temp=head;
              
        while( l1!=NULL && l2 !=NULL){
            if( l1->val <= l2-> val ){
                temp->next=l1;
                l1=l1->next;
            }
            else{
                temp->next=l2;
                l2=l2->next;
            }
            
            temp=temp->next;
        }
        
        if( l1!=NULL )
            temp->next=l1;
        else        // l2!=NULL 
            temp->next=l2;
            
        return head;
    }
};
